---
title:               "Habib 2017 snRNAseq Report"
author:
  - name:            "Emir Turkes"
  - name:            "Columbia University"
date:                '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography:        "../habib-2017-snRNAseq.bib"
biblio-style:        apalike
link-citations:      true
output:
  html_document:
    number_sections: true
    theme:           lumen
    toc:             true
    toc_depth:       2
    toc_float:
      collapsed:     false
      smooth_scroll: false

knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile,
      encoding = encoding,
      output_file = "../results/habib-2017-snRNAseq-report.html"
    )
  })
---

```{r, include = FALSE}
#    This file is part of habib-2017-snRNAseq.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(echo = TRUE)
```

<style type="text/css">

body{font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}

</style>

*This is a broad initial analysis that prepares and characterizes the data for use in other projects.*

The background for this data is as follows:

- dbGaP Accession: [phs000424](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v7.p2).
- Part of [GTEx](https://gtexportal.org/home/) and originally published in @habib_massively_2017.
- Archived frozen adult human post-mortem tissue.
- 19,550 nuclei, 1,683 genes, 2,187 transcripts.
- 3 PFC samples, 4 hippocampus samples from 5 donors.

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

# Final Results

**Read just this section for the final results of the analysis and a summary of the methods.**

# ~~~~ Breakdown of Methods ~~~~ {-}

The following sections breakdown the methods used to transform the contents in `Original Data` to those in `Final Results`.

# Original Data

# To Be Organized

```{r}
counts <- data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.umi_counts.txt.gz")),
  data.table = FALSE
)
rownames(counts) <- counts$V1
counts <- as.matrix(counts[ , -1])

keep <- grep("*PFC*", colnames(counts))
counts <- counts[ , keep]

part_cell_id <- sapply(colnames(counts), function(xx) strsplit(xx, "_")[[1]][1], USE.NAMES = FALSE)
col_dat <- data.frame(
  sample_name = colnames(counts),
  part_cell_id = part_cell_id,
  stringsAsFactors = FALSE,
  check.names = FALSE
)
```

```{r}
clust <- data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.clusters.txt.gz")),
  data.table = FALSE
)
names(clust) <- c("sample_name", "Habib_clusters")

clust <- clust[keep, ]

clust$Habib_clusters <- as.factor(clust$Habib_clusters)
cname = c(
  "exPFC1", "exPFC2", "exCA1", "exCA3", "GABA1", "GABA2", "exDG",
  "ASC1", "ASC2","ODC1", "ODC2", "OPC", "MG", "NSC", "END", rep(NA, 4)
)
ctype = c(
  "exPFC", "exPFC", "exCA1", "exCA3", "GABA", "GABA", "exDG",
  "ASC", "ASC", "ODC", "ODC", "OPC", "MG", "NSC", "END", rep(NA,4)
)
map_clust_name <- data.frame(
  Habib_clusters = as.factor(seq(19)),
  Habib_clusters_name = cname,
  Habib_cell_type = ctype,
  stringsAsFactors = FALSE,
  check.names = FALSE
)
clust <- merge(clust, map_clust_name, by = intersect(names(clust), names(map_clust_name)))
clust <- clust[match(col_dat$sample_name, clust$sample_name), ]
```

```{r}
df_tsne = data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.tsne.txt.gz")),
  data.table = FALSE
)
names(df_tsne) = c("sample_name", paste0("Habib_TSNE", 1:2))

df_tsne <- df_tsne[keep, ]

col_dat <- cbind(
  col_dat,
  clust[ , names(clust) != "sample_name"],
  df_tsne[ , names(df_tsne) != "sample_name"]
)
sce <- SingleCellExperiment::SingleCellExperiment(assays = list(counts = counts), colData = col_dat)
```

```{r}
if (file.exists("../results/gene_anno.rds")) {
  gene_anno <- readRDS("../results/gene_anno.rds")
} else {
  ensembl <- biomaRt::useEnsembl(biomart = "ensembl", GRCh = 37, dataset = "hsapiens_gene_ensembl")
  attr_string <- c(
    "hgnc_symbol", "ensembl_gene_id", "external_gene_name", "chromosome_name",
    "chromosome_name", "start_position", "end_position", "strand",
    "description", "percentage_gene_gc_content", "gene_biotype"
  )
  gene_anno <- biomaRt::getBM(
    attributes = attr_string, filters = "external_gene_name", values = rownames(sce), mart = ensembl
  )
  saveRDS(gene_anno, file = "../results/gene_anno.rds")
}
```

```{r}
remove_genes <- which(!gene_anno$external_gene_name %in% rownames(sce))
gene_anno <- gene_anno[-remove_genes, ]
```

```{r}
chromosomes <- c(1:22, "X", "Y", "MT")
gene_anno <- gene_anno[which(gene_anno$chromosome_name %in% chromosomes), ]
t1 <- table(gene_anno$external_gene_name)
t2 <- sort(t1[t1 > 1], decreasing = TRUE)
duplicates <- which(gene_anno$external_gene_name %in% names(t2))
gene_anno2 <- gene_anno[duplicates, ]
gene_anno2 <- gene_anno2[which(gene_anno2$hgnc_symbol == gene_anno2$external_gene_name), ]
gene_anno2 <- dplyr::distinct(gene_anno2, external_gene_name, .keep_all = TRUE)
gene_anno <- rbind(gene_anno[-duplicates, ], gene_anno2)
gene_missing <- setdiff(rownames(sce), gene_anno$external_gene_name)
keep_genes <- match(gene_anno$external_gene_name, rownames(sce))
sce <- sce[keep_genes, ]
SummarizedExperiment::rowData(sce) <- gene_anno
```

```{r}
bc_rank <- DropletUtils::barcodeRanks(BiocGenerics::counts(sce))
uniq <- !duplicated(bc_rank$rank)
par(mar = c(5, 4, 2, 1), bty = "n")
plot(
  bc_rank$rank[uniq],
  bc_rank$total[uniq],
  log = "xy",
  xlab = "Rank",
  ylab = "Total UMI count",
  cex = 0.5,
  cex.lab = 1.2
)
abline(h = bc_rank$inflection, col = "darkgreen", lty = 2, lwd = 2)
abline(h = bc_rank$knee, col = "dodgerblue", lty = 2, lwd = 2)
legend(
  "left",
  legend = c("Inflection", "Knee"),
  bty = "n",
  col = c("darkgreen", "dodgerblue"),
  lty = 2,
  cex = 1.2,
  lwd = 2
)
```

```{r}
ribo_genes <- read.table(
  "../results/ribo-genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE
)
is_mito <- which(SummarizedExperiment::rowData(sce)$chromosome_name == "MT")
is_ribo <- which(
  SummarizedExperiment::rowData(sce)$external_gene_name %in% ribo_genes$Approved.Symbol
)
sce <- scater::calculateQCMetrics(sce, feature_controls = list(Mt = is_mito, Ri = is_ribo))
par(mfrow = c(2, 2), mar = c(5, 4, 1, 1), bty = "n")
hist(
  log10(sce$total_counts), xlab = "log10(Library sizes)", main = "", 
  breaks = 20, col = "grey80", ylab = "Number of cells"
)
hist(
  log10(sce$total_features), xlab = "log10(# of expressed genes)", main = "",
  breaks = 20, col = "grey80", ylab = "Number of cells"
)
hist(
  sce$pct_counts_Ri, xlab = "Ribosome prop. (%)", ylab = "Number of cells",
  breaks = 40, main = "", col = "grey80"
)
hist(
  sce$pct_counts_Mt, xlab = "Mitochondrial prop. (%)", ylab = "Number of cells",
  breaks = 80, main = "", col = "grey80"
)
par(mfrow = c(2, 2), mar = c(5, 4, 1, 1), bty = "n")
smoothScatter(
  log10(sce$total_counts), log10(sce$total_features),
  xlab = "log10(Library sizes)", ylab = "log10(# of expressed genes)"
)
smoothScatter(
  log10(sce$total_counts), sce$pct_counts_Ri,
  xlab = "log10(Library sizes)", ylab = "Ribosome prop. (%)"
)
smoothScatter(
  log10(sce$total_counts), sce$pct_counts_Mt,
  xlab = "log10(Library sizes)", ylab = "Mitochondrial prop. (%)"
)
smoothScatter(
  sce$pct_counts_Ri,sce$pct_counts_Mt, xlab = "Ribosome prop. (%)", ylab = "Mitochondrial prop. (%)"
)
```

```{r}
libsize_drop <- scater::isOutlier(sce$total_counts, nmads = 3, type = "lower", log = TRUE)
feature_drop <- scater::isOutlier(sce$total_features_by_counts, nmads = 3, type = "lower", log = TRUE)
mito_drop <- scater::isOutlier(sce$pct_counts_Mt, nmads = 3, type = "higher")
ribo_drop <- scater::isOutlier(sce$pct_counts_Ri, nmads = 3, type = "higher")
keep <- !(libsize_drop | feature_drop | mito_drop | ribo_drop)
sce <- sce[ , keep]
```

```{r}
par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
  log10(SummarizedExperiment::rowData(sce)$mean_counts + 1e-6), col = "grey80",  main = "",
  breaks = 40, xlab = "log10(ave # of UMI + 1e-6)")
hist(
  log10(SummarizedExperiment::rowData(sce)$n_cells_counts + 1), col = "grey80", main = "",
  breaks = 40, xlab = "log10(# of expressed cells + 1)")
plot(
  log10(SummarizedExperiment::rowData(sce)$mean_counts + 1e-6),
  pch = 16,
  col = rgb(0, 0, 0, 0.4),
  log10(SummarizedExperiment::rowData(sce)$n_cells_counts + 1),
  xlab="log10(ave # of UMI + 1e-6)",
  ylab="log10(# of expressed cells + 1)"
)
tb1 <- table(SummarizedExperiment::rowData(sce)$n_cells_counts)
```
```{r}
names(SummarizedExperiment::rowData(sce))[
  names(SummarizedExperiment::rowData(sce)) == "strand"
] <- "strand_n"
n_genes <- colSums(BiocGenerics::counts(sce) >= 2)
n_genes <- colSums(BiocGenerics::counts(sce) >= 1)
n_cells <- rowSums(BiocGenerics::counts(sce) >= 2)
sce <- sce[which(n_cells >= 10), ]
```

```{r}
par(mar = c(5, 4, 1, 1))
od1 <- order(SummarizedExperiment::rowData(sce)$mean_counts, decreasing = TRUE)
barplot(
  SummarizedExperiment::rowData(sce)$mean_counts[od1[20:1]],
  las = 1,
  names.arg = SummarizedExperiment::rowData(sce)$hgnc_symbol[od1[20:1]],
  horiz = TRUE,
  cex.names = 0.8,
  cex.axis = 0.8,
  xlab = "ave # of UMI"
)
```
```{r}
clusters <- scran::quickCluster(sce, min.mean = 0.1, method = "igraph")
sce <- scran::computeSumFactors(sce, cluster = clusters, min.mean = 0.1)
sce <- sce[ , which(BiocGenerics::sizeFactors(sce) > 0.01)]
par(mfrow = c(1, 2), mar = c(5, 4, 2, 1), bty = "n")
smoothScatter(
  sce$total_counts, BiocGenerics::sizeFactors(sce), log = "xy",
  xlab = "total counts", ylab = "size factors"
)
plot(
  sce$total_counts, BiocGenerics::sizeFactors(sce), log = "xy", xlab = "total counts",
  ylab = "size factors", cex = 0.3, pch = 20, col = rgb(0.1, 0.2, 0.7, 0.3)
)
abline(h = 0.05)
sce <- sce[ , which(BiocGenerics::sizeFactors(sce) > 0.05)]
sce <- BiocGenerics::normalize(sce)
```

# Conclusion

This concludes the methods used in this analysis.
Below we simply write the data to disk for exploration in `Final Results` at the top of this report.

```{r}
devtools::session_info()
```

# References
