---
title:               "Habib 2017 snRNAseq Report"
author:
  - name:            "Emir Turkes"
  - name:            "Columbia University"
date:                '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography:        "../habib-2017-snRNAseq.bib"
biblio-style:        apalike
link-citations:      true
output:
  html_document:
    number_sections: true
    theme:           lumen
    toc:             true
    toc_depth:       2
    toc_float:
      collapsed:     false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding, output_file = "../results/habib-2017-snRNAseq-report.html")})
---

```{r, include = FALSE}
#    This file is part of habib-2017-snRNAseq.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(echo = TRUE)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This is a broad initial analysis that prepares and characterizes the data for use in other projects.*

The background for this data is as follows:

- dbGaP Accession: [phs000424](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v7.p2).
- Part of [GTEx](https://gtexportal.org/home/) and originally published in @habib_massively_2017.
- Archived frozen adult human post-mortem tissue.
- 19,550 nuclei, 1,683 genes, 2,187 transcripts.
- 3 PFC samples, 4 hippocampus samples from 5 donors.

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

# Final Results

**Read just this section for the final results of the analysis and a summary of the methods.**

# ~~~~ Breakdown of Methods ~~~~ {-}

The following top-level sections breakdown the methods used to perform the analysis.
We start by loading in any required packages for our analysis and setting some global variables.

```{r}
library(conflicted)
library(devtools)
library(data.table)
library(DT)
library(plyr)
library(SingleCellExperiment)
library(biomaRt)
```

```{r}
data_dir <- file.path(getwd(), "..", "data")
results_dir <- file.path(getwd(), "..", "results")
```

# Original Data {.tabset}

This section provides a brief look at the raw data before manipulation.

## Counts

```{r}
counts <- fread(file.path(data_dir, "GTEx_droncseq_hip_pcf.umi_counts.txt.gz"), data.table = FALSE)
datatable(counts[1:5, 1:3])
```

## Clusters

```{r}
clust <- fread(file.path(data_dir, "GTEx_droncseq_hip_pcf.clusters.txt.gz"), data.table = FALSE)
datatable(clust[1:5, ])
```

## t-SNE

```{r}
tsne = fread(file.path(data_dir, "../data/GTEx_droncseq_hip_pcf.tsne.txt.gz"), data.table = FALSE)
datatable(tsne[1:5, ])
```

# Preliminary Cleaning {.tabset}

Here we provide more informative labels to things and transform the data into more convenient formats for downstream analysis.

## Counts

```{r}
rownames(counts) <- counts$V1
counts <- as.matrix(counts[ , -1])
part_cell_id <- sapply(colnames(counts), function(xx) strsplit(xx, "_")[[1]][1], USE.NAMES = FALSE)
col_dat <- data.frame(
  "Sample" = colnames(counts),
  "Cell ID Stem" = part_cell_id,
  stringsAsFactors = FALSE,
  check.names = FALSE)
datatable(counts[1:5, 1:3])
```

## Clusters

```{r}
names(clust) <- c("Sample", "Habib Cluster")
clust$`Habib Cluster` <- as.factor(clust$`Habib Cluster`)

# Add cluster names from Supplementary Table 8 (nmeth.4407-S11.xlsx).
# The last four clusters lack labels.
cname = c(
  "exPFC1", "exPFC2", "exCA1", "exCA3", "GABA1", "GABA2", "exDG", "ASC1",
  "ASC2", "ODC1", "ODC2", "OPC", "MG", "NSC", "END", rep(NA,4))
ctype = c(
  "exPFC", "exPFC", "exCA1", "exCA3", "GABA", "GABA", "exDG", "ASC",
  "ASC", "ODC", "ODC", "OPC", "MG", "NSC", "END", rep(NA,4))

# Merge cluster names with clust.
map_clust_name <- data.frame(
  "Habib Cluster" = as.factor(seq(19)),
  "Habib Cluster Name" = cname,
  "Habib Cell Type" = ctype,
  stringsAsFactors = FALSE,
  check.names = FALSE)
clust <- join(clust, map_clust_name, by = intersect(names(clust), names(map_clust_name)))
clust <- clust[match(col_dat$Sample, clust$Sample), ]

datatable(clust[1:5, ])
```

## t-SNE

```{r}
names(tsne) = c("Sample", paste0("Habib tSNE ", 1:2))
col_dat <- cbind(
  col_dat, clust[ , names(clust) != "Sample"], tsne[ , names(tsne) != "Sample"])
datatable(tsne[1:5, ])
```

# SingleCellExperiment

The data is now sufficiently prepared to form a SingleCellExperiment (SCE) object.

```{r}
sce <- SingleCellExperiment(assays = list(counts = counts), colData = col_dat)
sce
```

# Gene Annotations

We add annotations based on the hg19 reference gnome, the same used in @habib_massively_2017.

```{r}
if (file.exists(file.path(results_dir, "gene_anno.rds"))) {
  gene_anno <- readRDS(file.path(results_dir, "../results/gene_anno.rds"))
} else {
  ensembl <- useEnsembl(biomart = "ensembl", GRCh = 37, dataset = "hsapiens_gene_ensembl")
  attr_string <- c(
    "HGNC Symbol", "Ensembl Gene ID", "External Gene Name", "Chromosome Name",
    "Start Position", "End Position", "Strand", "Description",
    "Percentage Gene GC Content", "Gene Biotype")
  gene_anno <- getBM(
    attributes = attr_string, filters = "external_gene_name", values = rownames(sce), mart = ensembl)
  saveRDS(gene_anno, file = file.path(results_dir, "gene_anno.rds"))}
```

# Conclusion

This is the concluding section of the document.
Here we write relevant results to disk, output the `session_info`, and create a bibliography for works cited.

```{r}
session_info()
```

# References
