---
title:               "Habib 2017 snRNAseq Report"
author:
  - name:            "Emir Turkes"
  - name:            "Columbia University"
date:                '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography:        "../habib-2017-snRNAseq.bib"
biblio-style:        apalike
link-citations:      true
output:
  html_document:
    number_sections: true
    theme:           lumen
    toc:             true
    toc_depth:       2
    toc_float:
      collapsed:     false
      smooth_scroll: false

knit:
  (function(inputFile, encoding) {rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = "../results/habib-2017-snRNAseq-report.html"
  )})
---

```{r, include = FALSE}
#    This file is part of habib-2017-snRNAseq.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(echo = TRUE)

library(BiocGenerics)
```

<style type="text/css">

body{ /* Normal  */
  font-size: 16px;
}
h1.title {
  font-size: 35px;
}
h1 { /* Header 1 */
  font-size: 24px;
}
h2 { /* Header 2 */
  font-size: 22px;
}
h3 { /* Header 3 */
  font-size: 20px;
}
.toc-content {
  padding-left: 0px;
  padding-right: 0px;
}
div.tocify {
  width: 100%;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
  padding-left: 25px;
  text-indent: 0;
}
div.main-container {
  max-width: none;
  width: 100%;
}

</style>

*This is a broad initial analysis that prepares and characterizes the data for use in other projects.*

The background for this data is as follows:

- dbGaP Accession: [phs000424](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000424.v7.p2).
- Part of [GTEx](https://gtexportal.org/home/) and originally published in @habib_massively_2017.
- Archived frozen adult human post-mortem tissue.
- 19,550 nuclei, 1,683 genes, 2,187 transcripts.
- 3 PFC samples, 4 hippocampus samples from 5 donors.

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

# Final Results

**Read just this section for the final results of the analysis and a summary of the methods.**

# ~~~~ Breakdown of Methods ~~~~ {-}

The following sections breakdown the methods used to transform the contents in `Original Data` to those in `Final Results`.

# Original Data

# To Be Organized

```{r}
counts <- data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.umi_counts.txt.gz")),
  data.table = FALSE
)
rownames(counts) <- counts$V1
counts <- as.matrix(counts[ , -1])
part_cell_id <- sapply(colnames(counts), function(xx) strsplit(xx, "_")[[1]][1], USE.NAMES = FALSE)
col_dat <- data.frame(
  "Sample Name" = colnames(counts),
  part_cell_id = part_cell_id,
  stringsAsFactors = FALSE,
  check.names = FALSE
)
```

```{r}
clust <- data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.clusters.txt.gz")),
  data.table = FALSE
)
names(clust) <- c("Sample Name", "Habib Clusters")
clust$"Habib Clusters" <- as.factor(clust$"Habib Clusters")
cname = c(
  "exPFC1", "exPFC2", "exCA1", "exCA3", "GABA1", "GABA2", "exDG",
  "ASC1", "ASC2","ODC1", "ODC2", "OPC", "MG", "NSC", "END", rep(NA, 4)
)
ctype = c(
  "exPFC", "exPFC", "exCA1", "exCA3", "GABA", "GABA", "exDG",
  "ASC", "ASC", "ODC", "ODC", "OPC", "MG", "NSC", "END", rep(NA,4)
)
map_clust_name <- data.frame(
  "Habib Clusters" = as.factor(seq(19)),
  "Habib Cluster Name" = cname,
  "Habib Cluster Type" = ctype,
  stringsAsFactors = FALSE,
  check.names = FALSE
)
clust <- merge(clust, map_clust_name, by = intersect(names(clust), names(map_clust_name)))
clust <- clust[match(col_dat$"Sample Name", clust$"Sample Name"), ]
```

```{r}
df_tsne = data.table::fread(
  cmd = sprintf("zcat < %s", file.path(getwd(), "../data/GTEx_droncseq_hip_pcf.tsne.txt.gz")),
  data.table = FALSE
)
names(df_tsne) = c("Sample Name", paste0("Habib t-SNE", 1:2))
col_dat <- cbind(
  col_dat,
  clust[ , names(clust) != "Sample Name"],
  df_tsne[ ,names(df_tsne) != "Sample Name"]
)
sce <- SingleCellExperiment::SingleCellExperiment(assays = list(counts = counts), colData = col_dat)
```

```{r}
if (file.exists("../results/gene_anno.rds")) {
  gene_anno <- readRDS("../results/gene_anno.rds")
} else {
  ensembl <- biomaRt::useEnsembl(biomart = "ensembl", GRCh = 37, dataset = "hsapiens_gene_ensembl")
  attr_string <- c(
    "hgnc_symbol", "ensembl_gene_id", "external_gene_name", "chromosome_name",
    "chromosome_name", "start_position", "end_position", "strand",
    "description", "percentage_gene_gc_content", "gene_biotype"
  )
  gene_anno <- biomaRt::getBM(
    attributes = attr_string, filters = "external_gene_name", values = rownames(sce), mart = ensembl
  )
  saveRDS(gene_anno, file = "../results/gene_anno.rds")
}
```

```{r}
remove_genes <- which(!gene_anno$external_gene_name %in% rownames(sce))
gene_anno <- gene_anno[-remove_genes, ]
```

```{r}
chromosomes <- c(1:22, "X", "Y", "MT")
gene_anno <- gene_anno[which(gene_anno$chromosome_name %in% chromosomes), ]
t1 <- table(gene_anno$external_gene_name)
t2 <- sort(t1[t1 > 1], decreasing = TRUE)
duplicates <- which(gene_anno$external_gene_name %in% names(t2))
gene_anno2 <- gene_anno[duplicates, ]
gene_anno2 <- gene_anno2[which(gene_anno2$hgnc_symbol == gene_anno2$external_gene_name), ]
gene_anno2 <- dplyr::distinct(gene_anno2, external_gene_name, .keep_all = TRUE)
gene_anno <- rbind(gene_anno[-duplicates, ], gene_anno2)
gene_missing <- setdiff(rownames(sce), gene_anno$external_gene_name)
keep_genes <- match(gene_anno$external_gene_name, rownames(sce))
sce <- sce[keep_genes, ]
SummarizedExperiment::rowData(sce) <- gene_anno
```

```{r}
bc_rank <- DropletUtils::barcodeRanks(counts(sce))
uniq <- !duplicated(bc_rank$rank)
plot(
  bc_rank$rank[uniq],
  bc_rank$total[uniq],
  log = "xy",
  xlab = "Rank",
  ylab = "Total UMI count",
  cex = 0.5,
  cex.lab = 1.2
)
abline(h = bc_rank$inflection, col = "darkgreen", lty = 2, lwd = 2)
abline(h = bc_rank$knee, col = "dodgerblue", lty = 2,lwd = 2)
legend(
  "left",
  legend = c("Inflection", "Knee"),
  bty = "n",
  col = c("darkgreen", "dodgerblue"),
  lty = 2,
  cex = 1.2,
  lwd = 2
)
set.seed(100)
e_out <- DropletUtils::emptyDrops(counts(sce))
tapply(e_out$Total, e_out$FDR, summary)
```

# Conclusion

This concludes the methods used in this analysis.
Below we simply write the data to disk for exploration in `Final Results` at the top of this report.

```{r}
devtools::session_info()
```

# References
